reconcile/utils/three_way_diff_strategy.py

```diff
@@ -0,0 +1,126 @@
+import base64
+import logging
+import re
+from collections.abc import Mapping
+from typing import Any
+
+import jsonpatch  # type: ignore
+from jsonpointer import resolve_pointer  # type: ignore
+
+from reconcile.utils.openshift_resource import QONTRACT_ANNOTATIONS
+from reconcile.utils.openshift_resource import OpenshiftResource as OR
+
+NORMALIZE_COMPARE_EXCLUDED_ATTRS = {
+    "creationTimestamp",
+    "resourceVersion",
+    "generation",
+    "selfLink",
+    "uid",
+    "fieldRef",
+    "managedFields",
+    "namespace",
+}
+
+
+def _normalize_secret(secret: OR) -> None:
+    body = secret.body
+    string_data = body.get("stringData")
+
+    if string_data:
+        data = body.get("data") or {}
+        body["data"] = data | {
+            k: base64.b64encode(str(v).encode()).decode("utf-8")
+            for k, v in string_data.items()
+        }
+        secret.body = {k: v for k, v in body.items() if k != "stringData"}
+
+
+def _normalize_noop(item: OR) -> None:
+    pass
+
+
+NORMALIZERS = {"Secret": _normalize_secret}
+
+
+def normalize_object(item: OR) -> OR:
+    # Remove K8s managed attributes not needed to compare objects
+    metadata = {
+        k: v
+        for k, v in item.body["metadata"].items()
+        if k not in NORMALIZE_COMPARE_EXCLUDED_ATTRS
+    }
+
+    n = OR(
+        body=item.body | {"metadata": metadata},
+        integration=item.integration,
+        integration_version=item.integration_version,
+        error_details=item.error_details,
+        caller_name=item.caller_name,
+        validate_k8s_object=False,
+    )
+
+    annotations = n.body.get("annotations", {})
+    metadata["annotations"] = {
+        k: v for k, v in annotations.items() if k not in QONTRACT_ANNOTATIONS
+    }
+
+    # Run normalizers on Kinds with special needs
+    NORMALIZERS.get(n.body["kind"], _normalize_noop)(n)
+    return n
+
+
+def check_cpu_mutation(current: str, desired: str) -> bool:
+    return OR.cpu_equal(current, desired)
+
+
+def is_a_valid_mutation(current: OR, desired: OR, patch: Mapping[str, Any]) -> bool:
+    pointer = patch["path"]
+    if re.match(r"/.*/(requests|limits)/cpu$", pointer):
+        current_value = resolve_pointer(current.body, pointer)
+        desired_value = patch["value"]
+        return check_cpu_mutation(current_value, desired_value)
+
+    return False
```

Comment:
Is this mean all non cpu changes are invalid?

Reply:
No. This means that the change in the patch has not been detected as a valid mutation and the value in the patch is not equal to the value pointed by the pointer in the current object. 

I added a method docstring to clarify this. If you have a batter naming or any idea I can consider it.

Reply:
Oh, I see, it's a custom equal check, since it's only used in valid changes filter, how about extract them all into one method?

```python
def is_valid_change(current: OR, desired: OR, patch: Mapping[str, Any]) -> bool:
    if patch["op"] not in ["add", "replace"]:
        return False

    pointer = patch["path"]
    if re.match(r"/.*/(requests|limits)/cpu$", pointer):
        current_value = resolve_pointer(current.body, pointer)
        desired_value = patch["value"]
        return not OR.cpu_equal(current_value, desired_value)
    return True
```

```python
    valid_changes = [
        item for item in patch.pathc if is_valid_change(current, desired, item)
    ]
```

btw that re expression can be extracted into a compiled modulo const to avoid recompile on runtime


```python
CPU_REGEX = re.compile(r"/.*/(requests|limits)/cpu$")
```

Reply:
Yup, I think this way is more clear. I maintained the concept of mutations though as I expect to add more cases and think that being explicit in the code about them is more clear.

Reply:
Oh, I finally understand `mutation` you mean they are the same value but different forms, I thought it means changed, maybe need a better word.

Reply:
`variant`?

Reply:
The CPU is just a case. AFAIK K8s use "mutation" for this. This code will manage all the mutations made by Mutating admission controllers (https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook)  and other controllers as well.

One case that I have in mind (and that's why I pass the desired object to the method) is related to the Skupper-managed objects. Services with Skupper annotations are deeply modified by the Skupper controller, even changing attributes in the "desired" service with Skupper-related attributes. 

For now, we are not using annotated services anywhere, but the option is there. With this approach, we can easily add more methods to control these behaviors if needed.

Reply:
TBH, I think this is splitting hairs a little bit. I think `is_cpu_mutation` is clear enough and there is also a comment in the code saying this is going to check mutations. 

As per the old code and checking a bit the current pods, memory does not get mutated the same way (there are pods with requests/memory = 1000Mi and are not transformed into 1Gi for example).

Reply:
I also feel it's clear enough, but maybe because of the context in this thread, not sure if someone read the code the first time will get this.

I remember I saw a loop once `1.5Gi` can't compare with `1536Mi` in memory, I have to specify it in `Mi` most time, and only integer value of `Gi` can work without loop.